static/app/components/comboBox/index.tsx

```diff
@@ -0,0 +1,413 @@
+import {
+  type Key,
+  useCallback,
+  useContext,
+  useEffect,
+  useLayoutEffect,
+  useMemo,
+  useRef,
+  useState,
+} from 'react';
+import isPropValid from '@emotion/is-prop-valid';
+import {useTheme} from '@emotion/react';
+import styled from '@emotion/styled';
+import {useComboBox} from '@react-aria/combobox';
+import {Item, Section} from '@react-stately/collections';
+import {type ComboBoxStateOptions, useComboBoxState} from '@react-stately/combobox';
+import omit from 'lodash/omit';
+
+import {SelectFilterContext} from 'sentry/components/compactSelect/list';
+import {ListBox} from 'sentry/components/compactSelect/listBox';
+import {
+  getDisabledOptions,
+  getEscapedKey,
+  getHiddenOptions,
+  getItemsWithKeys,
+} from 'sentry/components/compactSelect/utils';
+import Input from 'sentry/components/input';
+import LoadingIndicator from 'sentry/components/loadingIndicator';
+import {Overlay, PositionWrapper} from 'sentry/components/overlay';
+import {t} from 'sentry/locale';
+import {space} from 'sentry/styles/space';
+import mergeRefs from 'sentry/utils/mergeRefs';
+import type {FormSize} from 'sentry/utils/theme';
+import useOverlay from 'sentry/utils/useOverlay';
+
+import {SelectContext} from '../compactSelect/control';
+
+import type {
+  ComboBoxOption,
+  ComboBoxOptionOrSection,
+  ComboBoxOptionOrSectionWithKey,
+} from './types';
+
+interface ComboBoxProps<Value extends string>
+  extends ComboBoxStateOptions<ComboBoxOptionOrSection<Value>> {
+  'aria-label': string;
+  className?: string;
+  disabled?: boolean;
+  isLoading?: boolean;
+  size?: FormSize;
+  sizeLimit?: number;
+  sizeLimitMessage?: string;
+}
+
+function ComboBox<Value extends string>({
+  size = 'md',
+  className,
+  placeholder,
+  disabled,
+  isLoading,
+  sizeLimitMessage,
+  menuTrigger = 'focus',
+  ...props
+}: ComboBoxProps<Value>) {
```

Comment:
I'd also add a `leadingItems` and `trailingItems` prop, similar to what we have for `MenuListItem`. That way we can add things like leading search icon, trailing chevron, etc.
<img width="278" alt="Screenshot 2024-04-09 at 3 39 33 PM" src="https://github.com/getsentry/sentry/assets/44172267/b4ea4826-12b9-4e4d-929b-44e6f25e32ef">

Reply:
If you wanna make it really nice we can also use a composable structure like with `InputGroup`:
https://github.com/getsentry/sentry/blob/7e22a005f94b6316140495d2225ba99161440941/static/app/components/inputGroup.spec.tsx#L30-L35

Reply:
good idea! I created [an issue](https://github.com/getsentry/sentry/issues/68592) for it. Will do it in a follow up!